torcharrow/dtypes_core.py

```diff
@@ -0,0 +1,342 @@
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int8(nullable)
+
+
+@dataclass(frozen=True)
+class Int16(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "s"
+    arraycode: ty.ClassVar[str] = "h"
+    name: ty.ClassVar[str] = "int16"
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int16(nullable)
+
+
+@dataclass(frozen=True)
+class Int32(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "i"
+    arraycode: ty.ClassVar[str] = "i"
+    name: ty.ClassVar[str] = "int32"
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int32(nullable)
+
+
+@dataclass(frozen=True)
+class Int64(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "l"
+    arraycode: ty.ClassVar[str] = "l"
+    name: ty.ClassVar[str] = "int64"
+    default: ty.ClassVar[int] = 0
+
+    def constructor(self, nullable):
+        return Int64(nullable)
+
+
+# Not all Arrow types are supported. We don't have a backend to support unsigned
+# integer types right now so they are removed to not confuse users. Feel free to
+# add unsigned int types when we have a supporting backend.
+
+
+@dataclass(frozen=True)
+class Float32(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "f"
+    arraycode: ty.ClassVar[str] = "f"
+    name: ty.ClassVar[str] = "float32"
+    default: ty.ClassVar[float] = 0.0
+
+    def constructor(self, nullable):
+        return Float32(nullable)
+
+
+@dataclass(frozen=True)
+class Float64(Numeric):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "g"
+    arraycode: ty.ClassVar[str] = "d"
+    name: ty.ClassVar[str] = "float64"
+    default: ty.ClassVar[float] = 0.0
+
+    def constructor(self, nullable):
+        return Float64(nullable)
+
+
+@dataclass(frozen=True)
+class String(DType):
+    nullable: bool = False
+    typecode: ty.ClassVar[str] = "u"  # utf8 string (n byte)
+    arraycode: ty.ClassVar[str] = "w"  # wchar_t (2 byte)
+    name: ty.ClassVar[str] = "string"
+    default: ty.ClassVar[str] = ""
+
+    def constructor(self, nullable):
+        return String(nullable)
+
+
+@dataclass(frozen=True)
+class Map(DType):
+    key_dtype: DType
+    item_dtype: DType
+    nullable: bool = False
+    keys_sorted: bool = False
+    name: ty.ClassVar[str] = "Map"
+    typecode: ty.ClassVar[str] = "+m"
+    arraycode: ty.ClassVar[str] = ""
+
+    @property
+    def py_type(self):
+        return ty.Dict[self.key_dtype.py_type, self.item_dtype.py_type]
+
+    def constructor(self, nullable):
+        return Map(self.key_dtype, self.item_dtype, nullable)
+
+    def __str__(self):
+        nullable = ", nullable=" + str(self.nullable) if self.nullable else ""
+        return f"Map({self.key_dtype}, {self.item_dtype}{nullable})"
+
+    def default_value(self):
+        return {}
+
+
+@dataclass(frozen=True)
+class List(DType):
+    item_dtype: DType
```

Comment:
Here we use `item_dtype` and `fixed_size` while in PyArrow it's called `value_type` and `list_size`: https://arrow.apache.org/docs/python/generated/pyarrow.list_.html

 `fixed_size` vs. `list_size`  seems fine to me.  I guess `item` can be renamed to `value` :). (since `ListColumn` also contains `values` and `offsets`.

Reply:
or, `element_dtype` as per `Velox` and `Trino/Presto`: 

Velox: https://github.com/facebookincubator/velox/blob/ee631f54e9646e0442b577926d6bff4633e93401/velox/type/Type.cpp#L577-L580

Trino: https://github.com/trinodb/trino/blob/8d6b82e874ea7e79722bccacc0ebbca950b8908f/core/trino-spi/src/main/java/io/trino/spi/type/ArrayType.java#L86-L90